### PR TITLE
Synchronize SSH keys on login with LDAP + Fix SQLite deadlock on ldap ssh key deletion

### PR DESCRIPTION
**Pull request by zeripath**
This PR synchronises SSH keys on login with LDAP. 

It could do with a test from an LDAP user.

### CODE REVIEW

**Comment by zeripath** [comment]
Should fix #5545

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/login_source.go [hunk 1]
```diff
@@ -421,7 +430,19 @@ func LoginViaLDAP(user *User, login, password string, source *LoginSource, autoR
 		IsActive:    true,
 		IsAdmin:     sr.IsAdmin,
 	}
-	return user, CreateUser(user)
+
+	err := CreateUser(user)
+
+	if err == nil {
```
[the remark this answers is not in the document]
Comment (zeripath) [reply models/login_source.go]: Done.

File: models/login_source.go [hunk 2]
```diff
@@ -393,7 +393,16 @@ func LoginViaLDAP(user *User, login, password string, source *LoginSource, autoR
 		return nil, ErrUserNotExist{0, login, 0}
 	}
 
+	var isAttributeSSHPublicKeySet = len(strings.TrimSpace(source.LDAP().AttributeSSHPublicKey)) > 0
+
 	if !autoRegister {
+		if isAttributeSSHPublicKeySet {
+			needToUpdate := SynchronizeLdapSSHPublicKeys(user, source, sr.SSHPublicKey)
```
[the remark this answers is not in the document]
Comment (zeripath) [reply models/login_source.go]: done

**Comment by NomAnor** [comment]
I did a quick test and removed my user from gitea and then logged in. The ssh key was added.
Then I removed the key from my account and relogged and it was added again.
Also works with multiple keys. It looks to be doing everything I want. Thanks for the fast response.

Maybe add the 'sshPublicKey' attribute to the `log.Trace("Fetching attributes '%v', '%v', '%v', '%v', '%v' with filter %s and base %s", ...)` call

**Comment by zeripath** [comment]
@NomAnor cool.

It looks like the changes have broken some tests so I need to look at those and re-push. 
I'll look at changing the debug note too.

**Comment by zeripath** [comment]
OK, problem 1. is definitely a bug with the current implementation - Any sqlite installation will hang if they have to remove keys during a synchronisation.

**Comment by NomAnor** [comment]
I changed the full name of my account an relogged and it was not overwritten. Then I started the synchronization as administrator and it was changed. So the current behaviour of gitea seem to be that account details are only overwritten when the synchronization runs and initially filled out on first login without an existing account (Maybe document that relogging does currently not synchronize account information).

The current ssh key implementation does not adhear to this behaviour and synchronizes on every login. I think it would be better if all external authentication plugins and all account attributes work the same way.

If someone uses external account management then all information available should be overwritten so the account management is the definite source of information. From a user perspective I think it would be good that this synchronization also happens on every login (that might have performance issues on big installations; make it configurable?).  In this case the overwritten user settings should be disabled (like the username for external accounts).

I'm not in a corporate environment so their requirements might be different.

Another problem might occour if someone does not want to store ssh keys in the ldap directory. Then the keys are deleted by the synchronization because the sshPublicKey Attribute is absent. Only delete the keys if the sshPublicKey Attribute is present but empty?

**Comment by zeripath** [comment]
Hi @NomAnor. I think you're right - if you're going to have externally managed sshkeys then they should be externally managed. I could imagine a use-case whereby you want to allow people to add their own, but then if you go down that route I can imagine a use-case whereby you want you to be able to feed back in to the LDAP. And so on, and so on...

I guess the thing is, although this PR doesn't actually change the current policy it certainly makes it more apparent...

(I've fixed the problem with the test cases and the deadlock bug in sqlite now.)

File: models/user.go [hunk 3]
```diff
@@ -1421,35 +1421,37 @@ func deleteKeysMarkedForDeletion(keys []string) (bool, error) {
...
-			log.Warn("addLdapSSHPublicKeys[%s]: Skipping invalid LDAP Public SSH Key for user %s: %v", s.Name, usr.Name, sshKey)
+			log.Warn("AddLdapSSHPublicKeys[%s]: Skipping invalid LDAP Public SSH Key for user %s: %v", s.Name, usr.Name, sshKey)
 		}
 	}
 	return sshKeysNeedUpdate
 }
 
-func synchronizeLdapSSHPublicKeys(s *LoginSource, SSHPublicKeys []string, usr *User) bool {
+// SynchronizeLdapSSHPublicKeys updates a users public keys. Returns true if there are changes.
+func SynchronizeLdapSSHPublicKeys(usr *User, s *LoginSource, SSHPublicKeys []string) bool {
```
Comment (lafriks) [inline models/user.go]: Why make this function exported if it is not used outside models?

File: models/user.go [hunk 4]
```diff
@@ -1421,35 +1421,37 @@ func deleteKeysMarkedForDeletion(keys []string) (bool, error) {
 	return sshKeysNeedUpdate, nil
 }
 
-func addLdapSSHPublicKeys(s *LoginSource, usr *User, SSHPublicKeys []string) bool {
+// AddLdapSSHPublicKeys add a users public keys. Returns true if there are changes.
+func AddLdapSSHPublicKeys(usr *User, s *LoginSource, SSHPublicKeys []string) bool {
```
Comment (lafriks) [inline models/user.go]: Same as with sync function

File: models/user.go (hunk 3 above)
Comment (zeripath) [reply models/user.go]: You're right! I've changed the order of the arguments to be a little more consistent but I'll keep it private.

**Review by lafriks — APPROVED** [review]

**Review by techknowlogick — APPROVED** [review]

**Pull request merged**